solidity/contracts/converter/types/standard-pool/StandardPoolConverter.sol

```diff
@@ -0,0 +1,1362 @@
+        view
+        returns (uint256, uint256)
+    {
+        return readReserveBalances(1, 2);        
+    }
+
+    /**
+      * @dev converts a specific amount of source tokens to target tokens
+      * can only be called by the bancor network contract
+      *
+      * @param _sourceToken source ERC20 token
+      * @param _targetToken target ERC20 token
+      * @param _amount      amount of tokens to convert (in units of the source token)
+      * @param _trader      address of the caller who executed the conversion
+      * @param _beneficiary wallet to receive the conversion result
+      *
+      * @return amount of tokens received (in units of the target token)
+    */
+    function convert(IERC20Token _sourceToken, IERC20Token _targetToken, uint256 _amount, address _trader, address payable _beneficiary)
+        public
+        override
+        payable
+        protected
+        only(BANCOR_NETWORK)
+        returns (uint256)
+    {
+        // validate input
+        require(_sourceToken != _targetToken, "ERR_SAME_SOURCE_TARGET");
+
+        return doConvert(_sourceToken, _targetToken, _amount, _trader, _beneficiary);
+    }
+
+    /**
+      * @dev returns the conversion fee for a given target amount
+      *
+      * @param _targetAmount  target amount
+      *
+      * @return conversion fee
+    */
+    function calculateFee(uint256 _targetAmount) internal view returns (uint256) {
+        return _targetAmount.mul(conversionFee) / PPM_RESOLUTION;
+    }
+
+    /**
+      * @dev reads the stored reserve balance for a given reserve id
+      *
+      * @param _reserveId   reserve id
+    */
+    function readReserveBalance(uint256 _reserveId) internal view returns (uint256) {
+        return (__reserveBalances >> ((_reserveId - 1) * 128)) & MAX_UINT128;        
+    }
+
+    /**
+      * @dev reads the stored reserve balances
+      *
+      * @param _sourceId    source reserve id
+      * @param _targetId    target reserve id
+    */
+    function readReserveBalances(uint256 _sourceId, uint256 _targetId) internal view returns (uint256, uint256) {
+        require(_sourceId + _targetId == 3, "ERR_INVALID_RESERVES");
+        uint256 reserveBalancesLocal = __reserveBalances;
+        return (
+            (reserveBalancesLocal >> ((_sourceId - 1) * 128)) & MAX_UINT128,
+            (reserveBalancesLocal >> ((_targetId - 1) * 128)) & MAX_UINT128
+        );        
+    }
+
+    /**
+      * @dev writes the stored reserve balance for a given reserve id
+      *
+      * @param _reserveId       reserve id
+      * @param _reserveBalance  reserve balance
+    */
+    function writeReserveBalance(uint256 _reserveId, uint256 _reserveBalance) internal {
+        require(_reserveBalance <= MAX_UINT128, "ERR_RESERVE_BALANCE_OVERFLOW"); 
+        uint256 otherBalance = readReserveBalance(3 - _reserveId);
+        __reserveBalances = (_reserveBalance << ((_reserveId - 1) * 128)) | (otherBalance << ((2 - _reserveId) * 128));
```

Comment:
A utility function specifically for 128 bits and `MAX_UINT128`?
Because we also have similar code for 112 bits.
If we implement a generic function, then it will be much less gas-efficient.
If we implement a set of functions, then it will be a lot of duplicated code.

Reply:
let's close this, we already have this discussion above.

Reply:
OK, you can close all but one bitwise-related comment, while I'm trying to come up with the most appropriate solution.